metrics/prometheus/prometheus_test.go

```diff
@@ -0,0 +1,344 @@
+package metrics
+
+import (
+	"testing"
+	"time"
+
+	"github.com/prebid/prebid-cache/config"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/stretchr/testify/assert"
+)
+
+func createPrometheusMetricsForTesting() *PrometheusMetrics {
+	return CreatePrometheusMetrics(config.PrometheusMetrics{
+		Port:      8080,
+		Namespace: "prebid",
+		Subsystem: "cache",
+	})
+}
+
+func assertCounterVecValue(t *testing.T, description string, counterVec *prometheus.CounterVec, expected float64, labels prometheus.Labels) {
+	counter := counterVec.With(labels)
+	assertCounterValue(t, description, counter, expected)
+}
+
+func assertCounterValue(t *testing.T, description string, counter prometheus.Counter, expected float64) {
+	m := dto.Metric{}
+	counter.Write(&m)
+	actual := *m.GetCounter().Value
+
+	assert.Equal(t, expected, actual, description)
+}
+
+func assertGaugeValue(t *testing.T, description string, gauge prometheus.Gauge, expected float64) {
+	m := dto.Metric{}
+	gauge.Write(&m)
+	actual := *m.GetGauge().Value
+
+	assert.Equal(t, expected, actual, description)
+}
+
+func assertHistogram(t *testing.T, name string, histogram prometheus.Histogram, expectedCount uint64, expectedSum float64) {
+	m := dto.Metric{}
+	histogram.Write(&m)
+	actual := *m.GetHistogram()
+
+	assert.Equal(t, expectedCount, actual.GetSampleCount(), name+":count")
+	assert.Equal(t, expectedSum, actual.GetSampleSum(), name+":sum")
+}
+
+func TestPrometheusRequestStatusMetric(t *testing.T) {
+	m := createPrometheusMetricsForTesting()
+
+	type testCaseObject struct {
+		description      string
+		expDuration      float64
+		expRequestTotals float64
+		expRequestErrors float64
+		expBadRequests   float64
+		testCase         func(pm *PrometheusMetrics)
+	}
+
+	testGroups := map[*PrometheusRequestStatusMetric][]testCaseObject{
+		m.Puts: {
+			{
+				description: "Log put request duration",
+				testCase: func(pm *PrometheusMetrics) {
+					timeZero := time.Time{}
+					tenSeconds := time.Second * 10
+
+					now := timeZero.Add(tenSeconds)
+
+					pm.RecordPutDuration(&now)
+				},
+				expDuration:      9.223372036854776e+09,
```

Comment:
Where does this odd value come from?

Reply:
That's the value that the `prometheus.Histogram` holds. I'm not sure if the Histogram averages the quanttities before returning the value from the `Histogram.Value` function or if time.Second * 10 gets converted to that number from the beginning. That's the value with which the `assert.Equal` test could pass.

Reply:
This doesn't seem right. `pm.RecordPutDuration()` is expecting a duration, not a time 10 seconds from now. Shouldn't you be calling `pm.RecordPutDuration(&tenSeconds)` ... Your magic number might be the max possible duration?

Reply:
`pm.RecordPutDuration` expects a Time pointer. Unfortunately I named it `duration *time.Time` hence the confusion. Do you wnat me to change it to something else? Too `timeLapse *time.Time` maybe?
```
218 func (m *PrometheusMetrics) RecordPutDuration(duration *time.Time) {
219     m.Puts.Duration.Observe(time.Since(*duration).Seconds())
220 }
metrics/prometheus/prometheus.go
```
>... Your magic number might be the max possible duration?

I guess so, I'm honestly not sure.

Reply:
Ok, now I see. You may want to call the variable `startTime` or something like that. The function being tested looks at the difference between the current time and the time supplied. So in your test you are setting the clock 10 seconds in the future, and then the function tries to determine how much time has past since that future date. It seems that it takes the absolute value, and reports the difference in nanoseconds. Since by the time the function checks the time, it is not quite 10 seconds ahead of the current time, and thus we see "almost 10 seconds in nanoseconds" as the magic number. This is fragile, because if it takes more or less time to get from recording the time until recording the metric, this number will change. Probably should just check it is within some range.

Also should subtract 10 seconds instead of add, so that you are recording a time since some point in the past, rather than "time since some time in the future" since the latter is not well defined.

Reply:
Substituted `UpdateSince(ts time.Time)` for `Update(d time.Duration)` in metrics/influxdb/influxdb so the interface receives a `time.Duration` parameter instead of `time.Time` object which simplifies testing a lot. More inforation in a slack message I sent to the team's channel.